cmd/atlas/internal/cmdapi/diff.go

```diff
@@ -17,85 +24,284 @@ import (
+		diff, err = c.SchemaDiff(current.Schemas[0], desired.Schemas[0])
+		if err != nil {
+			return err
 		}
-		diff, err = toC.SchemaDiff(fromSchema, toSchema)
-		cobra.CheckErr(err)
 	}
-	p, err := toC.PlanChanges(ctx, "plan", diff)
-	cobra.CheckErr(err)
+	p, err := c.PlanChanges(ctx, "plan", diff)
+	if err != nil {
+		return err
+	}
 	if len(p.Changes) == 0 {
 		cmd.Println("Schemas are synced, no changes to be made.")
-		return
 	}
 	for _, c := range p.Changes {
 		if c.Comment != "" {
 			cmd.Println("--", strings.ToUpper(c.Comment[:1])+c.Comment[1:])
 		}
 		cmd.Println(c.Cmd)
 	}
+	return nil
+}
+
+type (
+	// stateReadCloser is a migrate.StateReader with an optional io.Closer.
+	stateReadCloser struct {
+		migrate.StateReader
+		io.Closer        // optional close function
+		Schema    string // in case we work on a single schema
+	}
+	// stateReaderConfig is given to stateReader.
+	stateReaderConfig struct {
+		urls    []string          // urls to create a migrate.StateReader from
+		dev     *sqlclient.Client // dev database connection
+		schemas []string          // schemas to work on
+	}
+)
+
+// stateReader returns a migrate.StateReader that reads the state from the given urls.
+func stateReader(ctx context.Context, config *stateReaderConfig) (*stateReadCloser, error) {
+	scheme, err := selectScheme(config.urls)
+	if err != nil {
+		return nil, err
+	}
+	parsed := make([]*url.URL, len(config.urls))
+	for i, u := range config.urls {
+		parsed[i], err = url.Parse(u)
+		if err != nil {
+			return nil, err
+		}
+	}
+	switch scheme {
+	// "file" scheme is valid for both migration directory and HCL paths.
+	case "file":
+		// Replaying a migration directory or evaluating an HCL file requires a dev connection.
+		if config.dev == nil {
+			return nil, errors.New("--dev-url cannot be empty")
+		}
+		if len(config.urls) > 1 {
+			// Consider urls being HCL paths.
+			return hclStateReader(ctx, config, parsed)
+		}
+		path := filepath.Join(parsed[0].Host, parsed[0].Path)
+		fi, err := os.Stat(path)
+		if err != nil {
+			return nil, err
+		}
+		if !fi.IsDir() {
+			// If there is only one url given, and it is a file, consider it an HCL path.
+			return hclStateReader(ctx, config, parsed)
+		}
+		files, err := os.ReadDir(path)
+		if err != nil {
+			return nil, err
+		}
+		// Check all files, if we find both HCL and SQL files, abort. Otherwise, proceed accordingly.
+		var hcl, sql bool
+		for _, f := range files {
+			if f.IsDir() {
+				// skip directories
+			}
+			ext := filepath.Ext(f.Name())
+			switch {
+			case hcl && ext == ".sql", sql && ext == ".hcl":
+				return nil, fmt.Errorf("ambiguos files: %q contains both SQL and HCL files", path)
+			case ext == ".hcl":
+				hcl = true
+			case ext == ".sql":
+				sql = true
+			}
+			// unknown extension, we don't care
+		}
+		switch {
+		case hcl:
+			return hclStateReader(ctx, config, parsed)
+		case !sql:
+			return nil, fmt.Errorf("%q contains neither SQL nor HCL files", path)
+		default:
```

Comment:
```suggestion
		default:
			return nil, fmt.Errorf("%q contains neither SQL nor HCL files", path)
		case sql:
```

I'll suggest "switching" the clauses.